src/main/scala/it/unibo/authsim/library/dsl/cryptography/algorithm/hash/HashFunction.scala

```diff
@@ -0,0 +1,120 @@
+package it.unibo.authsim.library.dsl.cryptography.algorithm.hash
+
+import com.google.common.hash.Hashing
+import com.google.common.io.BaseEncoding
+import it.unibo.authsim.library.dsl.cryptography.algorithm.CryptographicAlgorithm
+import org.apache.commons.codec.binary.Hex
+import org.apache.commons.codec.digest.DigestUtils
+
+import java.nio.charset.StandardCharsets
+
+/**
+ * Trait representing the hash function
+ */
+trait HashFunction extends CryptographicAlgorithm:
+  /**
+   * Method that perform the hash conversion: it converts the input value into a string created according
+   * to the hash algorithm chosen for the conversion.
+   *
+   * @param value             input value to transfrom
+   * @tparam A                type value of the input
+   * @return                  a string representing the hashed input value
+   */
+  def hash[A](value: A): String
+
+  /**
+   * Setter for the salt value
+   *
+   * @param salt              a new salt to use during the hash transformation
+   * @tparam A                type value of the input
+   */
+  def _salt[A](salt:A): Unit
+
+
+/**
+ * Companion object of the HashFunction trait
+ */
+object HashFunction:
+  import it.unibo.authsim.library.dsl.cryptography.util.ImplicitConversion._
+
+  /**
+   *  Class representing a basic implementation of the HashFunction algorithm
+   */
+  abstract class BasicHash extends HashFunction:
+    /**
+     * Protected value representing the salt
+     */
+    protected var _salt: String= ""
```

Comment:
Could `_salt` be an `Option[String]` rather than just `String`?

If not, how do you differentiate between an empty-string salt and a not-set salt (they may have the same behavior or represent the same thing)?